### PR TITLE
New Player Tools landing page, with stored sessions

### PR DESCRIPTION
**Pull request by ussjohnjay**
Created a new player tools landing page (`playertools.tsx`). It's functionally similar to the previous one, but reworks a lot of the internals to match the structure of the existing profile page and to make it easier to maintain going forward. All components have been moved inside the landing page, with a new component created for the unneeded items tool (`unneededitems.tsx`) and a new component that's exclusively dedicated to the voyage calculator (`voyagecalculator_iap.tsx`).

The top menu now points to the new player tools page and the address will be `datacore.app/playertools`. The existing page for player tools and the component for the voyage calculator remain in place until the new landing is fully reviewed, after which we can leave them up for legacy or redirect them to the new landing page. Other pages that point to the player tools may need to be updated.

All player tools now work with the same customized `playerData` variation (generated by `prepareProfileData`) and future components should use that instead of any other versions of player data for consistency.

A stripped version of player data (the same format that is used when sharing profiles) is now saved in session storage, so that a user will not have to re-import the player.json file if they click away from the player tools. A timestamp of the import is now shown on the landing page, with a button to allow users to update their player data. This player data only resides in session until the user closes the browser.

To do: copy the export button from the profile page over to the landing page, and maybe copy the 'Update now' button from the landing page to the profile page.

Import/upload form changes:
Renamed a few variables to be more clear with the new upload functionality.
The URL of the remote player data is now stored as a variable, to ensure that all links properly point to the same client API number.

Voyage calculator changes:
The voyage calculator now properly considers frozen crew when that option is toggled.
The excludes form now only shows crew that are on your active roster by default. When the consider frozen crew option is checked, the excludes shows all of your crew accordingly. The listed event crew also change depending on the toggle.
Reworded the options to say 'consider' instead of 'include', for clarity.

Unneeded items changes:
Moved unneeded items tab to after the items list, its placement here seems more logical.
The tool now shows the number of unneeded items for each group.

Top menu change: Removed the `content` property from popup, as that was triggering a react warning and the popup wasn't being shown anyway.

### CODE REVIEW

File: src/components/topmenu.tsx [hunk 1]
```diff
@@ -37,7 +37,7 @@ class TopMenu extends PureComponent<TopMenuProps, TopMenuState> {
 								<Dropdown.Item onClick={() => navigate('/bb')}>Text only</Dropdown.Item>
 							</Dropdown.Menu>
 						</Dropdown>
-						<Menu.Item onClick={() => navigate('/voyage')}>Player tools</Menu.Item>
+						<Menu.Item onClick={() => navigate('/playertools')}>Player tools</Menu.Item>
```
Comment (ineffyble) [inline src/components/topmenu.tsx:40]: Can we make a redirect component from /voyage? A bunch of people will have that bookmarked, and I am guessing some external sites link directly there.

**Comment by ineffyble** [comment]
@ussjohnjay Are you able to point at the lines of the voyage calculator code that reflect your new changes? Don't want to spend too long reviewing code that's literally just been copy-pasted from elsewhere.

**Comment by ineffyble** [comment]
You should be able to make life easier with sessionStorage by implementing this pattern: https://www.darrenlester.com/blog/syncing-react-state-and-session-storage

That should basically let your state be initialised from sessionStorage, and then all changes to state (for relevant state items) automatically flow back into it.

File: src/components/voyagecalculator_iap.tsx [hunk 2]
```diff
@@ -0,0 +1,388 @@
...
+
+		return consideredShips;
+	}
+
+	_bonusCrewForCurrentEvent(eventData: any[], crewlist: any[]): BonusCrew | undefined {
+		let result = new BonusCrew();
+
+		if (eventData && eventData.length > 0) {
+			let activeEvent = eventData
+			  .filter((ev) => (ev.seconds_to_end > 0))
```
Comment (ineffyble) [inline src/components/voyagecalculator_iap.tsx:278]: While we're making changes here, we should ignore events over a day away.

```suggestion
			  .filter((ev) => (ev.seconds_to_end > 0 && ev.seconds_to_start < 86400))
```

**Comment by ussjohnjay** [comment]
> 
> 
> @ussjohnjay Are you able to point at the lines of the voyage calculator code that reflect your new changes? Don't want to spend too long reviewing code that's literally just been copy-pasted from elsewhere.

It's all a bit different now, though to be clear I've only changed how your full roster is generated before it's passed on to the voyage calculator, which I didn't touch at all. I hadn't planned on doing even this much, but it seemed only logical to make the component compatible with the `preparedProfileData` variant of `playerData`, at the same time as I tried to fix #15. Both meant we would have had to rewrite the roster generation part anyway.

Specifically here are the methods worth looking at: `componentDidMount`: the old version spent a lot of lines doing stuff for the Unneeded Items tool, which I've now separated into its own component, and the only real thing it did for the voyage calculator here was populate the excludes list with all your crew. This new one first makes `preparedProfileData` ships and crew compatible with the existing voyage calculator; `preparedProfileData` crew do not have some properties that the voyage calculator will be looking for (e.g. `crew.skills`), so we have to put them back in. At the same time, we create two lists that are used to populate the excludes: one for all your crew and another for only your non-frozen crew (the default view, but one you can toggle now).

`_bestVoyageShip` and `_bonusCrewForCurrentEvent` are copied here from `voyageutils.ts`, and similarly adapted to use `preparedProfileData` instead of `playerData`.

Finally `_packVoyageOptions` now uses the crew roster we generated in `componentDidMount`, whereas before it used `playerData` crew, which on its own does not include frozen crew, hence the bug.

**Comment by flash82** [comment]
We already use the `localforage` package to keep the content of the crew search field stored.

https://github.com/stt-datacore/website/blob/30429ea7dff4fb0f8d5683d0a904e79eedb308a1/src/components/searchabletable.tsx#L56

**Comment by ussjohnjay** [comment]
> 
> 
> We already use the `localforage` package to keep the content of the crew search field stored.

Thanks for the tip! I looked into this a little and saw that localforage doesn't do sessions on its own. Apparently there's a custom driver for it, but that might be overly complicated when our usage here is so limited. But I'll mess around with it as well as the pattern that effy referenced earlier. Syncing state and session does sound like a good idea to try, at least.

This did get me thinking, though. Maybe we give users the option to keep player data in localStorage, too. We'll continue to use session by default but maybe change the update button (or even the timestamp span) to a dropdown with options like, "Update Now," "Remember Player Data" (which converts it from session to localStorage), and a command to explicitly "Clear Player Data".

This is a quick mockup of the timestamp turned into a dropdown:
Screen Shot 2021-02-20 at 12 19 49c

This menu might not be the best place for it, but we should also add profile management options somewhere on the tools page, like "Automatically Update" and "Stop Sharing". Not sure if the functionality for either of those exists yet.

**Comment by flash82** [comment]
> I looked into this a little and saw that localforage doesn't do sessions on its own.

If we store the player data in sessionStorage, we should do the same for the search field content because that one should be even more transitional.

Does it matter where the data is stored though? The way I see it, you already store the last import date (`let dtImported = new Date();`) and could easily use that to a) discard any previously saved data or  b) give a visible stale warning to the user.

**Comment by ussjohnjay** [comment]
That's a fair question. I personally like session because localStorage lives forever. That could be a privacy concern, in addition to issues with stale data. So we'd additionally have to tell users to remember to clear their player data if they don't want it stored forever, whereas session just does it automatically unless users opt in into using localStorage.

As with everything, I could be overthinking this.

**Comment by flash82** [comment]
I see your point about privacy, that's definitely a concern.

On the other hand, what's the expected lifetime of useful player data versus the length of a browser session? If we say 24 hours for the former, then I'd know from my own use-case that I open and close my browser multiple times during that period.

[localforage-cache](https://github.com/LuudJanssen/localforage-cache) is a localforage wrapper, which would continue to let us use localStorage but (on the next user visit) automatically removes stale data based after whichever timeout we decide to set. I also see the argument that this isn't a heavy use case and both a library and wrapper would be overkill.

**Comment by ussjohnjay** [comment]
After spending some time testing several different player data files and different combinations of options, and fixing any problems I could find, I feel pretty good about the voyage calculator and unneeded items tools now. They're working as expected or, at least in the case of the voyage calculator, working the same as the previous version. (There are some data files that seem to confound the voyage calculator, as #18 points out, no matter which version I use.)

The only thing that's left to lock down is the session behavior, and then add redirects from any old tools links. Oh, and shoot, re-reading the comments here now, I meant to add the `&& ev.seconds_to_start < 86400` event checker in this commit. I'll make sure that's done next.

**Comment by ineffyble** [comment]
>  I meant to add the `&& ev.seconds_to_start < 86400` event checker in this commit. I'll make sure that's done next.

I included that as a suggestion, meaning from this page you can accept it to have GitHub automatically commit the change :)

**Comment by Kirkman** [comment]
So glad to see this. As a user, it has always puzzled me that this app doesn't store user data in localstorage by default. Copy-and-pasting the giant JSON file every time you want to use player tools is annoying.

I can understand that there might be privacy concerns. I think you deal with those by having clear, easy-to-find buttons to delete the user data.

**Comment by AlexCPU** [comment]
> So glad to see this. As a user, it has always puzzled me that this app doesn't store user data in localstorage by default. Copy-and-pasting the giant JSON file every time you want to use player tools is annoying.
> 
> I can understand that there might be privacy concerns. I think you deal with those by having clear, easy-to-find buttons to delete the user data.

The reason it's been this way is that the voyage calculator tools (which most people use) are only correct for the current JSON.  If we used an old one, we'd end up recommending the wrong ship and crew as the traits and skills would be incorrect.

**Comment by ineffyble** [comment]
Yep, and we should ensure the voyage calculator refuses to operate if the stored data is too old.

**Comment by ussjohnjay** [comment]
My intention is to eventually allow the voyage calculator to work with a few manual inputs when it detects stale data. The only thing that's absolutely required for the calculator to work are the primary and secondary skills. The traits would be optional. Each matched crew trait only adds about a minute to the voyage, so the calculator is going to give you roughly the same result without traits because it's much more reliant on skill numbers than traits anyway. (The ship trait, while optional, would be a recommended input, because a matched trait there would add ~7 minutes by itself; that's worth the 1 extra click.)

Of course there's the problem of crew rosters being outdated, but that really only messes with the recommendation on the margins. Unless you've just added a lot of new voyage-ready crew, it's more likely that an outdated player data would actually underestimate the voyage run time, which is an acceptable error. (And even then we shouldn't be responsible for players not liking their recommendations because they ignored any warnings we give about using old data.)

I'm also hopeful that one day we can precalculate event crew, because that data isn't user-specific and the voyage calculator would then be able to use the latest event data (and not potentially stale player data) when it pre-excludes crew that are more useful in events than on voyages.

**Comment by ineffyble** [comment]
@ussjohnjay I'd love to see this merged soon, anything we can do to support you with final bits?

**Comment by ussjohnjay** [comment]
I'm still not quite sure about the presentation, but I think everything works as expected now. A few tests from anyone willing to import/upload their player data would go a long way to confirm that this is ready to go. And I've decided to keep the session storage pretty basic for now. Feedback from more users will help determine which direction to take after that.

My latest commit changes the `voyage.tsx` page to a notice that the player tools address has changed, as I thought that might be more appropriate right now than an automatic redirect. (Though wouldn't it be neat if we could eventually redirect /voyage and /crewretrieval to their respective tabs?) If you have a better way to handle the address change, please let me know.

There are a couple of pages with /voyage links that need to be updated, but I chose not to do them here to limit the number of files changed in this one merge as much as possible.

**Comment by ineffyble** [comment]
Copying from Discord to ensure it's not lost:

@[UFP] USS John Jay there's some merge conflicts between your branch and master, I think changes to voyage.tsx that can't be merged automatically because you moved files around. I went to sort them out, and merged in one more pending PR for voyage calculator changes, but the extent of your changes is greater than I realised and I'm not entirely certain where I should be applying the changes
So I might need to ask you to fix that up so all the in-limbo changes are applied in the right place, and then I'll get it merged ASAP to ensure there's no future issues
The two commits I know aren't reflected in your branch are:
* New icon/modal while voyage calculation in progress (what I just merged)
* The metadata/SEO changes I made to all pages

But not certain there aren't others

**Comment by ussjohnjay** [comment]
Closing this, as the core functionality is now in #120

**Pull request closed without merging**